### DISCUSSION TITLE
can the Redball demo run on the gazebo env?

### DISCUSSION THREAD

**Discussion by jimingda** [robotology/community#658] [Q&A] [answered]
Hi, 

can i run the red ball demo on the gazebo rather than real icub?

Thanks.

**Comment by pattacini** [comment]
Hi @jimingda 

A couple of years ago, we prepared a specific test for the red-ball-demo that relies on Gazebo.
Our tests are available at https://github.com/robotology/icub-tests.

In detail, you can find more on this test at https://github.com/robotology/icub-tests/pull/58, where there's also a video showcasing the main operations.

**Reply by jimingda** [reply]
Thanks @pattacini 

Actually what i want to do is to integrate the stereo camera of two eyes in my own driving application: fixed the relative position of steer model to icub and use the inverse kinematic as well as cartesiancontrol. 

Now i want to change the control mode, firstly recognize the steering wheel model by stereo image, sense the position of the steer to eyes and then control the hand or fingers to grasp it.

My current problem is that i don't know how to integrate the camera in gazebo, so i ask for help here,  maybe i can refer to the red ball demo in gazebo.

**Reply by jimingda** [reply]
Hi @pattacini   

Since i came across the same error as before, how can i add the parameter `-s libgazebo_yarp_clock.so`,  i didn't find the gzserver command

```
icub@icubsrv:~/workspace/icub-tests/suites$ robottestingframework-testrunner --verbose --suite demoRedBall-icubSim.xml
Loading demoRedBall-icubSim.xml
[INFO] |yarp.os.Port|/icubsrv/robottestingframework-testrunner/105506/clock:i| Port /icubsrv/robottestingframework-testrunner/105506/clock:i active at tcp://192.168.101.112:10002/
[INFO] |yarp.os.Network| Success: port-to-port persistent connection added.
[ERROR] |yarp.os.NetworkClock| Cannot find time port "/clock" or a time topic "/clock@"
[INFO] |yarp.os.Time| Waiting for clock server to start broadcasting data ...

```

**Reply by jimingda** [reply]
I check the `icubsim-demoRedBall-fixture.xml` included in the `demoRedBall-icubSim.xml` on the system, both of two files has the parameter  `-s libgazebo_yarp_clock.so` in the command `gzserver`.

```
  1 <?xml version="1.0" encoding="UTF-8"?>
  2 
  3 <suite name="Demo Red Ball Test Suite">
  4     <description>Testing Red Ball Demo functionalities</description>
  5     <environment>--robotname icubSim</environment>
  6     <fixture param="--fixture icubsim-demoRedBall-fixture.xml"> yarpmanager </fixture>
  7 
  8     <test type="dll" param="--context demoRedBall --from config-test.ini"> DemoRedBallTest </test>
  9 </suite>
```

```
icub@icubsrv:~/workspace/icub-tests/suites$ locate icubsim-demoRedBall-fixture.xml
/exports/code/robotology-superbuild/src/icub-tests/suites/fixtures/icubsim-demoRedBall-fixture.xml
/home/icub/workspace/icub-tests/suites/fixtures/icubsim-demoRedBall-fixture.xml
```
```
  8     <module>
  9         <name>gzserver</name>
 10         <parameters>-s libgazebo_yarp_clock.so grasp-ball-gazebo.sdf</parameters>
 11         <node>localhost</node>
 12     </module>
```

**Reply by pattacini** [reply, marked as answer]
I reckon that the red ball demo is not a good example in that it makes use of a monocular approach. Only one camera is used and the depth (thus, the 3D position of the ball) is triangulated knowing the geometry of the ball.

Cameras in Gazebo are handled in the corresponding URDF model. For example, look at `iCubGazeboV2_5_visuomanip`:
- https://github.com/robotology/icub-models/blob/master/iCub_manual/robots/iCubGazeboV2_5_visuomanip/model.urdf#L2972-L3042

This model is used in [`icub-gazebo-grasping-sandbox`](https://github.com/robotology/icub-gazebo-grasping-sandbox), where a stereo approach is adopted.

---

Try to remove the env variable `YARP_CLOCK=/clock`. It's an easier solution.
